### ISSUE TITLE
Point GO rules at Noctua-generated GPADs and generate a report

### ISSUE THREAD

**Issue by vanaukenk** [geneontology/pipeline#220]
As part of the GAF 2.2 release, we'll need to repair Noctua-generated gene product-to-term relations, particularly for cellular component.

To do this, we'll need to:

1.  Run the existing GO rules over the Noctua-generated GPADs so that we repair relations, as needed
2. Produce a report so we know what's been repaired

As an aside, we need to understand what will happen with, for example, the SynGO annotations and potentially communicate with that group about the result.

Also, we will need to update the existing GO-CAM models and the Noctua UI accordingly so that our underlying annotation data is as up-to-date as possible wrt our annotation guidelines.

**Comment by vanaukenk** [comment]
Thanks @kltm and @dustine32 
Are these changes coming down the pipe with the latest pipeline run?
I've seen on skyhook that a run started early this morning, but will wait to do more QC until that run is completed.

**Comment by kltm** [comment]
@vanaukenk Noting that this now seems to be functioning. For example:
http://snapshot.geneontology.org/reports/
http://snapshot.geneontology.org/reports/noctua_pr-report.html

**Comment by kltm** [comment]
Talking to @vanaukenk and @dougli1sqrd 
We'll update the final layout on skyhook so that:
`noctua_wb-src.gpad.gz (minerva output)`
`noctua_wb.gpad.gz (final post-processed/upgraded file)`
No other intermediates.

**Comment by ukemi** [comment]
This is good news for us because we can test the loads as soon as it is available with no additional code tweaks. When will this be?

**Comment by ukemi** [comment]
Something weird is going on in the pipeline though. I see this in the 'valid' file and it isn't in this model.:
MGI	MGI:101782	contributes_to	GO:0015081	PMID:8107805	ECO:0000266	UniProtKB:P37089		20200109	MGI		contributor=http://orcid.org/0000-0003-3394-9805|noctua-model-id=gomodel:5df932e000000204|model-state=production

But if you look at this model, this annotation isn't there.
This line in the noctua_mgi.gpad file:
MGI	MGI:101782	contributes_to	GO:0015081	PMID:8107805	ECO:0000266	UniProtKB:P37089		20200605	MGI		contributor=http://orcid.org/0000-0001-7476-6306|model-state=production|noctua-model-id=gomodel:5ed1aab700000290
Looks correct, at least with respect to exactly what the model says.

**Comment by ukemi** [comment]
Scratch this. Looking at this file again, the two lines are now correct. I have no idea what we were seeing.

**Comment by ukemi** [comment]
I figured out what happened. It was a spreadsheet sorting error where all the columns failed to sort. Crisis averted.

**Comment by kltm** [comment]
@ukemi `snapshot` stopped and restarted with the code updates for the file laocation. We should have results on skyhook in the afternoon (PDT) sometime tomorrow, with files to snapshot in the late evening.

**Comment by ukemi** [comment]
@vanaukenk path forward? I suggest that we wait until this pushes through the system. The MGI pipeline will run on the weekend and the corrections will be incorporated into our gaf2.2 and GPAD2.0 files. If I check the products once these are live in snapshot, I should only require a brief look on Monday to make sure there wasn't anything dramatic. Then I think we will be ready to give the green light on all of this.

**Comment by vanaukenk** [comment]
@ukemi @kltm 
I am okay with waiting to check the full round trip; we can confirm on the manager's call.

A proposed timeline:
1) Once snapshot finishes today (3/23), @ukemi and @vanaukenk will again check the src and output Noctua GPADs
2) Once the MGI pipeline runs again over the weekend, @ukemi will check the MGI src files on Monday (3/29)and make sure all is okay.
3) If everything looks fine at MGI, we will flip the switch next Monday (3/29).

**Comment by kltm** [comment]
@vanaukenk @ukemi @pgaudet Not a judgement--I just pull the lever here--but waiting until next week will greatly increase the chance that we do not make a "March" release. Again, it's just a number, but I wanted to make sure everybody was okay with that potential.

**Comment by vanaukenk** [comment]
Thanks @kltm 
Let's weigh the pros and cons on tomorrow's managers call and decide then what is the best way forward.

**Comment by vanaukenk** [comment]
@kltm @dougli1sqrd @ukemi 
I checked skyhook this morning and I see the Noctua GPAD src files but not the filtered files yet.  Is this just a timing issue and they'll be coming along later?
Thx.

**Comment by kltm** [comment]
@vanaukenk It looks like there is a hiccup in the pipeline that may affect things going through today (geneontology/pipeline#220). We'll see if we can get ahead of it.

**Comment by kltm** [comment]
@vanaukenk Okay, we had a hangup in late pipeline, so things were not flushed out to snapshot.geneontology.org, but we do have results at:
http://skyhook.berkeleybop.org/snapshot/products/annotations/
with reports at:
http://skyhook.berkeleybop.org/snapshot/reports/

**Comment by ukemi** [comment]
This morning I checked the Noctua files that are in products on snapshot. I reviewed CC, BP and MF annotations and saw that the annotations generated were correct and were being 'fixed' correctly. The only issue I saw with this file:
noctua_mgi.gpad.gz
was that the header was incomplete. I believe it should have a date. We have not yet loaded this file into MGI, but I suspect our pipeline will catch up over the next few days. 
BOTTOM LINE: all looks good from this end!

**Comment by vanaukenk** [comment]
Hi - So far, I've checked the Noctua WB src and production files here for CC annotations:  http://snapshot.geneontology.org/products/annotations/index.html

I only found one issue that affects two CC annotations.  

There are two CC annotations in the source file to an obsolete GO term, GO:0000784, that are repaired to GO:0000781 which is correct.

However, it looks like the correct gp2term relation is subsequently not applied, as the two annotations to GO:0000781, still have a gp2term relation of 'part of' instead of being repaired to 'located in'.

I don't think it's worth holding up the pipeline for this (and will update the existing annotations in my model - perfect use case for ART!), but perhaps we should check the order of operations here?

@kltm @dustine32 @dougli1sqrd

**Comment by kltm** [comment]
@dougli1sqrd Will need to answer for the order of operations.

That said, as it sounds like the "noctua_" files seem to be good enough, I'm ready to trigger (or wait) as desired.

**Comment by kltm** [comment]
Talking to @dougli1sqrd , we'll try to setup a quick slot w/ @vanaukenk to discuss where this falls on the bug-documentation spectrum, what a "fix" might look like, etc.

**Comment by dougli1sqrd** [comment]
@vanaukenk do you have the specific lines and which file you see this occurring in? Also curious if this shows up in the reports (as both an upgrade of obsolete term and "upgrade" the relation? Maybe a live chat will help

**Comment by ukemi** [comment]
This morning @loricorbani modified the MGI import scripts so that the CC annotations with the located_in qualifier are legally imported into MGI and will be put in our gaf. They should appear in our gaf tomorrow morning. This means that we can trigger the GO release on Tuesday night. We did notice one oddity that there was a new provider in this file (GO). I tracked down the very few annotations and noticed that they were attributed to a GOC:@@@ curator. I replaced them with my orcID. I think this should solve the issue. If not, we can add GO as a valid contributor and they will be allowed to go through.

**Comment by kltm** [comment]
@ukemi Sounds good! Give the thumbs up from your end here and we'll go ahead and trigger the next `release` run.

**Comment by ukemi** [comment]
Hi @kltm. I just checked our production file and as near as I can tell, the missing annotations are all back. Let's turn on the valve for the pipeline.

**Comment by ukemi** [comment]
Just checking the QC reports because I am trying to track down why my annotations are going missing. It turns out that many of my annotations are being filtered out by rule #1 (http://snapshot.geneontology.org/reports/noctua_mgi-report.html). I'm not sure that these are true errors. The error checking needs to be fixed or the offending annotation extensions should not be added to the annotations in this manner.

ERROR - Syntax error in annotation extension field: extensions should be relation(curie) (involved_in_positive_regulation_of(GO:0003674)) -- MGI MGI:1929608 part_of GO:0005667 PMID:19911006 ECO:0000314 20190625 MGI part_of(CL:0000169),part_of(GO:0005634),capable_of_part_of(GO:0030968),capable_of_part_of(GO:2001243),enables(GO:0003700),involved_in(GO:0030968),involved_in(GO:2001243),involved_in_positive_regulation_of(GO:0003674),involved_in_negative_regulation_of(GO:0097193) noctua-model-id=gomodel:5ce58dde00001215|contributor=http://orcid.org/0000-0001-7476-6306|model-state=production

**Comment by ukemi** [comment]
And what's going on with rule 28? It looks like it is flagging everything.

**Comment by kltm** [comment]
Pinging @dougli1sqrd @dustine32 .

**Comment by dustine32** [comment]
Looks like it's flagging extension relations that ontobio [doesn't currently recognize](https://github.com/biolink/ontobio/blob/0ec3aa6fea9d4492a9873a4b9b394c4866f741b6/ontobio/rdfgen/relations.py#L27-L110):

involved_in_positive_regulation_of
starts
starts_with
ends_with

I think we just need to add these to the ontobio `relations.py` file. @dougli1sqrd Does this sound right?

**Comment by dustine32** [comment]
For rule 28, this is a bug that still needs fixing. The rule test expects the `aspect` field to be set but, since this is a GPAD, the field is blank. Though luckily it's a warning, so it shouldn't cause any annotations to be ejected.

**Comment by ukemi** [comment]
Thanks @dustine32. I thought I had seen this in another file that you sent to me, but wanted to be certain it was tracked. Sorry for re-reporting.

**Comment by pgaudet** [comment]
Hi, 

I am curious - we use 'involved_in_positive_regulation_of'; do we use 'involved in regulation of' and ''involved in negative regulation of'? I thought we used 'regulates'. 

Also, the list here contains many deprecated relations 

https://github.com/biolink/ontobio/blob/0ec3aa6fea9d4492a9873a4b9b394c4866f741b6/ontobio/rdfgen/relations.py#L27-L110

such as '"regulates activity of"; "regulates expression of"; isn't GOREL and /or RO checked for validity? Or else what is the process to deprecate a relation? 

Thanks, Pascale

**Comment by dougli1sqrd** [comment]
Just to build off what Dustin has said, rule 28 is harmless in this context because it's reading GPAD which has no aspect, and then makes one from the GO Term. But then because it's also written as gpad no aspect is used either. This is just an artifact of how annotations are modeled inside ontobio and how gorules are for the most part source agnostic.

And yeah, for the relation, ontobio hasn't been taught that relation yet. We can add that.

@pgaudet yeah, that list of relations is manually updated at the moment as ontobio hasn't used any relations ontology to determine relation validity. That's on the radar, and has been for some time, but this was the solution in the meantime to get things working. There was some push some time back where we wanted to get MGI annotations to pass through ontobio and so I added relations that MGI was using in the extensions that ontobio didn't recognize. I just looked up the label in GOREL or in RO to find the relation. If it was there I added it. If not I kept it as an error and @ukemi did fix a handful of those as well.

Lastly, gorule-0000001 is strictly for "does this parse as an annotation" and the only requirement (for extension relations) is just that the syntax is correct, and that each relation label has a corresponding URI. It's purely a data thing: can this line of text fit into what we think of as an annotation shaped hole. Any other processing of deprecation, correct usage, etc would be somewhere else down the line, whether another GO Rule or some other kind of processing by someone else.

**Comment by ukemi** [comment]
We are in a bit of a chicken and egg situation here. I think for now, adding the relations as 'valid' will work for getting them through, but we still need to take a detailed, formal look at the GPAD output from Noctua-generated annotations and perhaps cull the list of AEs that are generated. Once we do that, we should synchronize these tools. I think at the end of the day, we will decide that we might not want these extensions.  ping @vanaukenk

**Comment by ukemi** [comment]
I just made a GPAD where I added the filtered annotations back that I can use for the GOC demo that @deustp01 and I want to do for the meeting. But we should still fix this problem as it is filtering out useful information.

**Comment by kltm** [comment]
@ukemi Okay, I think that this ticket is done--we have input and output GPADs and are generating a report.

But, from the conversation above, it sounds like there is another ticket to be opened and developed? Looking above it sounds like there should maybe be a new ticket like:

```
The relations:
involved_in_positive_regulation_of
starts
starts_with
ends_with
should be added to ontobio for the purposes of rule 28
```

Is this correct?

**Comment by ukemi** [comment]
Hi @kltm. This is a very good question. Unfortunately I'm not certain about exactly what ontobio is meant to cover. In an ideal world, we would tidy up the GPAD outputs and these relations would not be in them. But, if ontobio is meant to cover everything then these relations are not 'wrong', so they should be added.

**Comment by dustine32** [comment]
@kltm Correct, I just made this ticket https://github.com/biolink/ontobio/issues/548. Thanks for the reminder and it should be real quick to implement!

**Comment by ukemi** [comment]
Thanks @dustine32. At least this will fix the immediate issue.

**Comment by kltm** [comment]
Okay, we have the new issue and reported exist, so I'll close this one out.
